sys/hoon.hoon

```diff
@@ -4264,296 +4270,477 @@
   |=  {{tab/@ edg/@} tac/tank}  ^-  wall
   (~(win re tac) tab edg)
 ::
+::  This library includes `plume`, the actual pretty printing logic,
+::  and a handful of utilities for constructing plums.
+::
+::  Generally, you'll just use `plume` like this:
+::
+::    ~(plume tall plum)  ::  Pretty print `plum` in tall mode.
+::    ~(plume falt plum)  ::  Pretty print `plum` in wide mode.
+::
+::  There is probably no reason to look at the utility routines unless
+::  you are writing something to generate `plum`s.
+::
+:: |^  :*  plume=plume
+::         fixed=fixed
+::         tall-fixed=tall-fixed
+::         tall-running=tall-running
+::         rune=rune
+::         simple-wide=simple-wide
+::         subtree=subtree
+::         sexp=sexp
+::      ==
+::
+::  This is the pretty-printer.  Use the `flat` arm to render a plum
+::  into a single line and use the `tall` arm to get a nice multi-line
+::  rendering that switches to wide mode if there's enough space.
+::
+::  For details about how this works and what exactly it does in various
+::  cases, take a look at the docs for `plum`, `plumfmt`, and at the
+::  docs on the arms of this door.
+::
 ++  plume
   |_  =plum
   ::
+  ::  An line, indented by `indent` spaces.
+  ::
+  +$  line  [indent=@ud text=tape]
+  ::
+  ::  An sequence of indented lines.
+  ::
+  +$  block  (list line)
+  ::
   ::  +flat: print as a single line
   ::
   ++  flat
+    ^-  wain
     text:linear
   ::
   ::  +tall: print as multiple lines
   ::
   ++  tall
     ^-  wain
     %+  turn  window
-    |=  [indent=@ud text=tape]
+    |=  line
     (crip (runt [indent ' '] text))
   ::
   ::  +adjust: adjust lines to right
   ::
   ++  adjust
-    |=  [tab=@ud =(list [length=@ud text=tape])]
-    (turn list |=([@ud tape] [(add tab +<-) +<+]))
-  ::  
+    |=  [tab=@ud =block]  ^-  ^block
+    (turn block |=([@ud tape] [(add tab +<-) +<+]))
+  ::
+  ::  Prepend `n` spaces to a tape.
+  ::
+  ++  prepend-spaces
+    |=  [n=@ t=tape]  ^-  tape
+    (runt [n ' '] t)
+  ::
+  ::  Append an element to the end of a list.
+  ::
+  ++  snoc
+    |*  {a/(list) b/*}
+    (weld a ^+(a [b]~))
+  ::
   ::  +window: print as list of tabbed lines
   ::
   ++  window
-    ^-  (list [indent=@ud text=tape])
-    ::  memoize for random access
-    ::
-    ~+  
-    ::  trivial text
-    ::
-    ?@  plum  [0 (trip plum)]~
+    ^-  block
+    ~+                                                  ::  for random access
+    ?@  plum  [0 (trip plum)]~                          ::  trivial text
     ?-  -.plum
-      ::  %|: text wrap
-      ::
-      %|  ::  wrapping stub, should wrap text to 40 characters
-          ::
-          [0 +:linear]~
       ::
-      ::  %&: text tree
+      ::  %para: Line-wrappable paragraph. This is a stub; it should
+      ::  wrap text to 40 characters.
+      ::
+      %para
```

Comment:
We usually reverse the indentation for `?-`s, like this:
```
?-    -.plum
    %para
  [0 +:linear]~
::
    %sbrk
  =/  sub  kid.plum
  ...
==
```

Reply:
That's true.